IDCs: various code optimisations

matthias-kleiner opened the pull request:
- creation of status map for pads which are dead or delivering high IDC values using IDC0 map
- applying pad status map on FLPs when calculating 1D-IDCs
- fix race condition when using multithreading during calculation of IDC1
- using the timestamp information instead of TFid when writing IDCs to CCDB
- printing of debug informations
- do not use weights during averaging/grouping in case no weights were set
- setting of the z-axis when drawing factorized IDCs

matthias-kleiner committed:
IDCs: Code optimisations

- creation of status map for pads which are dead or delivering high IDC
  values using IDC0 map
- applying pad status map on FLPs when calculating 1D-IDCs
- fix race condition when using multithreading during calculation of
  IDC1
- using the timestamp information instead of TFid when writing IDCs to CCDB
- printing of debug informations
- do not use weights during averaging/grouping in case no weights were set
- setting of the z-axis when drawing factorized IDCs

wiechula commented on Detectors/TPC/workflow/include/TPCWorkflow/TPCFourierTransformAggregatorSpec.h:
```diff
@@ -109,20 +107,22 @@ class TPCFourierTransformAggregatorSpec : public o2::framework::Task
   int mProcessedTFs{0};                         ///< number of processed time frames to keep track of when the writing to CCDB will be done
 
   /// \return returns TF of current processed data
   uint32_t getCurrentTF(o2::framework::ProcessingContext& pc) const { return o2::framework::DataRefUtils::getHeader<o2::header::DataHeader*>(pc.inputs().getFirstValid(true))->tfCounter; }
 
+  uint64_t getCurrentTimeStamp(o2::framework::ProcessingContext& pc) const { return DataRefUtils::getHeader<DataProcessingHeader*>(pc.inputs().getFirstValid(true))->creation; }
```
Perhaps these two functions can be put in https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/TPC/workflow/include/TPCWorkflow/ProcessingHelpers.h
And I would prefer to rename `getCurrentTimeStamp` to `getCreationTime`.

matthias-kleiner replied:
Thanks for the comment. I prefer to change this and some minor things (update the extracted pad status map only when it changed) in another PR. Can it then be merged now?

wiechula reviewed (changes requested):
Just minor things.

wiechula reviewed (dismissed):
Fine for me. Minor things can be done later.

matthias-kleiner committed:
updating extracted outlier map only when it changed

- simplifying time stamp range when writing IDCs to CCDB
- CalDet, CalArray adding oeprator to check for equality
- added TF and creation time to ProcessingHelpers

wiechula commented:
@davidrohr @shahor02 , seems the error is unrelated. Can this be merged?

wiechula commented on Detectors/TPC/base/include/TPCBase/CalArray.h:
```diff
@@ -267,6 +270,18 @@ inline const CalArray<T>& CalArray<T>::operator/=(const CalArray<T>& other)
+    LOG(error) << "pad subset type of the objects it not compatible";
+    return false;
+  }
+  bool isSame = mData == other.mData;
```
Readability braces.
```suggestion
  bool isSame = (mData == other.mData);
```

matthias-kleiner replied:
Thanks, should I fix it in this PR?
Otherwise I would create a new PR with some other minor changes and fix it there...

wiechula replied:
I thinks this can be fixed in a separate PR. It would be good to get these modification in soon, I think.

shahor02 replied:
The FST fails due to the alien-token problem on the builder, but this PR cannot affect the FST, shall I merge this?

matthias-kleiner replied:
From my side it can be merged

matthias-kleiner committed:
CalArray: Readability braces

- IDC factorization: fixing wrong memory access for last region

wiechula reviewed (approved):
@shahor02 , from my side this can be merged.
